### PR TITLE
feat(server): Enable RBAC for SSO. Closes #3525

### LINKED ISSUE #3525: Role-based Access Controls for SSO
# Summary

Currently any user that logs in using SSO escalates to the `argo-server` service account. We would like other options.

# Motivation

TBD

# Proposal

Either:

* **Casbin style** similar to Argo CD. 
* **Client mode style** i.e. map groups to a service account.

See:

* #1813

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [ ] ~~I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.~~
* [x] My builds are green. Try syncing with master if they are not. 

Closes #3525

### CODE REVIEW

**Comment by alexec** [comment]
@jessesuen this is as discussed, so should be no surprise.

File: docs/argo-server-sso.md [hunk 1]
```diff
@@ -41,3 +41,69 @@ kubectl delete secret sso
...
+sso:
+  # ...
+  scopes:
+   - groups
+  rbac:
+    enabled: true
+```
+
+!!! Note
+    Not all OIDC provider support the groups scope. Please speak to your provider about their options.
```
[the remark this answers is not in the document]
Comment (alexec) [reply docs/argo-server-sso.md:66]: Yes. Groups are not mandatory for this feature, but I expect you need some way to identify the groups a user is in it to be useful.

File: server/auth/gatekeeper.go [hunk 2]
```diff
@@ -112,42 +126,128 @@ func getAuthHeader(md metadata.MD) string {
...
+		return "", fmt.Errorf("failed to get service account: %w", err)
+	}
+	if len(serviceAccount.Secrets) == 0 {
+		return "", fmt.Errorf("expected at least one secret for SSO RBAC service account: %w", err)
+	}
+	secret, err := s.kubeClient.CoreV1().Secrets(s.namespace).Get(serviceAccount.Secrets[0].Name, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get service account secret: %w", err)
+	}
+	return "Bearer " + string(secret.Data["token"]), nil
```
Comment (jessesuen) [inline server/auth/gatekeeper.go:236]: As you're probably aware, this is inefficient, but acceptable for MVP.

File: docs/argo-server-sso.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexec) [reply docs/argo-server-sso.md:66]: Do we know of any providers that use something other than groups?

**Review by jessesuen — APPROVED** [review]

File: docs/argo-server-sso.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jessesuen) [reply docs/argo-server-sso.md:66]: IIRC, I think azure used a different scope. Let me check Argo CD

[the remark this answers is not in the document]
Comment (jessesuen) [reply docs/argo-server-sso.md:66]: It as AWS. See:
https://github.com/argoproj/argo-cd/issues/1471

**Pull request merged**